community/kernel/src/main/java/org/neo4j/kernel/impl/store/NeoStore.java

```diff
@@ -365,62 +363,78 @@ public int getRecordSize()
-            if ( newTrailerOffset > trailerOffset )
+            if ( pagedFile.getLastPageId() != -1 )
             {
-                assert trailerBuffer != null;
-                channel.position( newTrailerOffset );
-                channel.write( trailerBuffer );
+                try ( PageCursor pageCursor = pagedFile.io( 0, PagedFile.PF_EXCLUSIVE_LOCK ) )
+                {
+                    if ( pageCursor.next() )
+                    {
+                        byte[] trailerData = UTF8.encode( buildTypeDescriptorAndVersion( TYPE_DESCRIPTOR ) );
+                        byte[] allData = new byte[pagedFile.pageSize()];
+                        do
+                        {
+                            pageCursor.getBytes( allData );
+                        }
+                        while ( pageCursor.shouldRetry() );
+                        int trailerOffset = StoreVersionCheck.findTrailerPattern( allData, trailerData );
+                        int recordOffset = RECORD_SIZE * position.id;
+                        if ( recordOffset < trailerOffset )
+                        {
+                            // We're overwriting a record, get the previous value
+                            long record;
+                            byte inUse;
+                            do
+                            {
+                                pageCursor.setOffset( recordOffset );
+                                inUse = pageCursor.getByte();
+                                record = pageCursor.getLong();
+                            }
+                            while ( pageCursor.shouldRetry() );
+                            if ( inUse == Record.IN_USE.byteValue() )
+                            {
+                                previousValue = record;
+                            }
+                        }
+                        // Write the value
+                        do
+                        {
+                            pageCursor.setOffset( recordOffset );
+                            pageCursor.putByte( Record.IN_USE.byteValue() );
+                            pageCursor.putLong( value );
+                        }
+                        while ( pageCursor.shouldRetry() );
+
+                        // Append the trailer if we cut it off previously
+                        int newTrailerOffset = recordOffset + RECORD_SIZE;
+                        if ( newTrailerOffset > trailerOffset )
+                        {
+                            do
+                            {
+                                pageCursor.setOffset( newTrailerOffset );
+                                pageCursor.putBytes( trailerData );
+                            }
+                            while ( pageCursor.shouldRetry() );
+                        }
+                    }
+                }
             }
-            return previous;
         }
+        return previousValue;
```

Comment:
Are we sure this is sufficiently tested e.g. when the trailer is in the middle of where records could be written, or right after, or on page id 1? Old pre-2.2 stores would have their trailer added right after the last record, and post-2.2 stores most likely have it added at the beginning of page 1.